packages/encryption/src/lib/encryption.ts

```diff
@@ -182,49 +135,41 @@ export const decryptFromIpfs = async (
       errorCode: LIT_ERROR.INVALID_PARAM_TYPE.name,
     });
 
-  try {
-    const metadata: EncryptToIpfsPayload = await (
-      await fetch(`https://gateway.pinata.cloud/ipfs/${ipfsCid}`)
-    ).json();
-    if (metadata.dataType === 'string') {
-      return decryptToString(
-        {
-          accessControlConditions: metadata.accessControlConditions,
-          evmContractConditions: metadata.evmContractConditions,
-          solRpcConditions: metadata.solRpcConditions,
-          unifiedAccessControlConditions:
-            metadata.unifiedAccessControlConditions,
-          ciphertext: metadata.ciphertext,
-          dataToEncryptHash: metadata.dataToEncryptHash,
-          chain: metadata.chain,
-          authSig,
-          sessionSigs,
-        },
-        litNodeClient
-      );
-    } else {
-      return decryptToFile(
-        {
-          accessControlConditions: metadata.accessControlConditions,
-          evmContractConditions: metadata.evmContractConditions,
-          solRpcConditions: metadata.solRpcConditions,
-          unifiedAccessControlConditions:
-            metadata.unifiedAccessControlConditions,
-          ciphertext: metadata.ciphertext,
-          dataToEncryptHash: metadata.dataToEncryptHash,
-          chain: metadata.chain,
-          authSig,
-          sessionSigs,
-        },
-        litNodeClient
-      );
-    }
-  } catch (e) {
-    return throwError({
-      message: 'Unable to fetch or decrypt from IPFS',
-      errorKind: LIT_ERROR.INVALID_ARGUMENT_EXCEPTION.kind,
-      errorCode: LIT_ERROR.INVALID_ARGUMENT_EXCEPTION.name,
-    });
+  // FIXME: The return type of this function is inferrable based on the value of `params.dataType`
```

Comment:
Here the type sig if you want fix this:
```
export function decryptFromJson<T extends DecryptFromJsonProps>(
  params: T
): T extends { parsedJsonData: { dataType: 'file' } }
  ? ReturnType<typeof decryptToFile>
  : T extends { parsedJsonData: { dataType: 'string' } }
  ? ReturnType<typeof decryptToString>
  : never;
export function decryptFromJson<T extends DecryptFromJsonProps>(
  params: T
) {
```

Reply:
You need to overload the function to get the inferred return to work with a discriminated union, so you can't use an anonymous function as far as I know 🤷

Reply:
💖 That's awesome. I'll add that now.
Named functions are better for stack traces anyway :+1: I copypasta'd some existing code without thinking... have been trying to migrate to using named functions as I go along, thanks for the reminder

Reply:
Got it to work without the overload and with ReturnType<> :tada:.  Just had to use Promise<> to wrap the entire thing - guessing the weirdness was because the other methods were marked `async` and the error message we saw was misleading :+1: